File: OpenList-Frontend/src/utils/certificate.ts
import { r } from "."
import { PResp } from "~/types"

// 证书数据类型
export interface Certificate {
  id: number
  name: string
  type: "user" | "node"
  status: "pending" | "valid" | "expiring" | "revoked" | "rejected"
  owner: string
  owner_id?: number
  content?: string
  expiration_date: string | Date
  issued_date: string | Date
  created_at: string
  updated_at: string
}

// 证书申请数据类型
export interface CertificateRequest {
  id: number
  user_name: string
  user_id?: number
  type: "user" | "node"
  status: "pending" | "valid" | "rejected"
  reason: string
  approved_by?: string
  approved_at?: string
  rejected_by?: string
  rejected_at?: string
  rejected_reason?: string
  created_at: string
  updated_at: string
}

// 管理员接口
export const getCertificates = (page = 1, per_page = 10): Promise<PResp<Certificate[]>> => {
  return r.get("/admin/certificate/list", {
    params: { page, per_page } as { page: number; per_page: number }
  })
}

export const getCertificateRequests = (page = 1, per_page = 10): Promise<PResp<CertificateRequest[]>> => {
  return r.get("/admin/certificate/requests", {
    params: { page, per_page } as { page: number; per_page: number }
  })
}

export const createCertificate = (data: Partial<Certificate>): Promise<PResp<Certificate>> => {
  return r.post("/admin/certificate/create", data)
}

export const createCertificateRequest = (data: Partial<CertificateRequest>): Promise<PResp<CertificateRequest>> => {
  return r.post("/admin/certificate/request", data)
}

export const updateCertificate = (id: number, data: Partial<Certificate>): Promise<PResp<Certificate>> => {
  return r.put(`/admin/certificate/update/${id}`, data)
}

export const deleteCertificate = (id: number): Promise<PResp<any>> => {
  return r.delete(`/admin/certificate/delete/${id}`)
}

export const revokeCertificate = (id: number): Promise<PResp<any>> => {
  return r.post(`/admin/certificate/revoke/${id}`)
}

export const approveCertificateRequest = (id: number, data: { approved_by: string }): Promise<PResp<Certificate>> => {
  return r.post(`/admin/certificate/request/approve/${id}`, data)
}

export const rejectCertificateRequest = (id: number, data: { rejected_by: string; rejected_reason?: string }): Promise<PResp<CertificateRequest>> => {
  return r.post(`/admin/certificate/request/reject/${id}`, data)
}

export const downloadCertificate = (id: number): Promise<Blob> => {
  return r.get(`/admin/certificate/download/${id}`, {
    responseType: 'blob' as 'blob'
  })
}

// 租户接口
export const getTenantCertificates = (): Promise<PResp<Certificate[]>> => {
  return r.get("/tenant/certificates")
}

export const getTenantCertificate = (): Promise<PResp<Certificate | null>> => {
  return r.get("/tenant/certificate")
}

export const getTenantCertificateRequests = (): Promise<PResp<CertificateRequest[]>> => {
  return r.get("/tenant/certificate/requests")
}

export const createTenantCertificateRequest = (data: Pick<CertificateRequest, "type" | "reason">): Promise<PResp<CertificateRequest>> => {
  return r.post("/tenant/certificate/request", data)
}

export const downloadTenantCertificate = (): Promise<Blob> => {
  return r.get("/tenant/certificate/download", {
    responseType: 'blob' as 'blob'
  })
}

File: OpenList-main/internal/model/certificate.go
package model

import (
"time"
"gorm.io/gorm"
)

// CertificateType 证书类型
type CertificateType string

const (
CertificateTypeUser CertificateType = "user" // 用户证书
CertificateTypeNode CertificateType = "node" // 节点证书
)

// CertificateStatus 证书状态
type CertificateStatus string

const (
CertificateStatusPending  CertificateStatus = "pending"   // 待审批
CertificateStatusValid    CertificateStatus = "valid"     // 有效
CertificateStatusExpiring CertificateStatus = "expiring"  // 即将过期
CertificateStatusRevoked  CertificateStatus = "revoked"   // 已吊销
CertificateStatusRejected CertificateStatus = "rejected"  // 已拒绝
)

// Certificate 证书实体
type Certificate struct {
	ID             uint              `json:"id" gorm:"primaryKey"`                        // unique key
	Name           string            `json:"name" gorm:"not null;index"`                 // 证书名称
	Type           CertificateType   `json:"type" gorm:"not null;index"`                 // 证书类型
	Status         CertificateStatus `json:"status" gorm:"not null;index"`               // 证书状态
	Owner          string            `json:"owner" gorm:"not null;index"`                // 证书所有者(用户名)
	OwnerID        uint              `json:"owner_id" gorm:"index"`                      // 证书所有者ID
	Content        string            `json:"content" gorm:"type:text"`                   // 证书内容(PEM格式)
	IssuedDate     time.Time         `json:"issued_date"`                                // 颁发日期
	ExpirationDate time.Time         `json:"expiration_date"`                            // 过期日期
	CreatedAt      time.Time         `json:"created_at"`
	UpdatedAt      time.Time         `json:"updated_at"`
	DeletedAt      gorm.DeletedAt    `gorm:"index" json:"deleted_at,omitempty"`
}

// CertificateRequest 证书申请实体
type CertificateRequest struct {
	ID             uint              `json:"id" gorm:"primaryKey"`                        // unique key
	UserName       string            `json:"user_name" gorm:"not null;index"`            // 申请人用户名
	UserID         uint              `json:"user_id" gorm:"index"`                       // 申请人用户ID
	Type           CertificateType   `json:"type" gorm:"not null"`                       // 申请证书类型
	Status         CertificateStatus `json:"status" gorm:"not null;index"`               // 申请状态
	Reason         string            `json:"reason" gorm:"type:text"`                    // 申请理由
	ApprovedBy     string            `json:"approved_by,omitempty"`                      // 审批人
	ApprovedAt     *time.Time        `json:"approved_at,omitempty"`                      // 审批时间
	RejectedBy     string            `json:"rejected_by,omitempty"`                      // 拒绝人
	RejectedAt     *time.Time        `json:"rejected_at,omitempty"`                      // 拒绝时间
	RejectedReason string            `json:"rejected_reason,omitempty" gorm:"type:text"` // 拒绝理由
	CreatedAt      time.Time         `json:"created_at"`
	UpdatedAt      time.Time         `json:"updated_at"`
	DeletedAt      gorm.DeletedAt    `gorm:"index" json:"deleted_at,omitempty"`
}

// IsValid 检查证书是否有效
func (c *Certificate) IsValid() bool {
return c.Status == CertificateStatusValid || c.Status == CertificateStatusExpiring
}

// IsExpired 检查证书是否过期
func (c *Certificate) IsExpired() bool {
return time.Now().After(c.ExpirationDate)
}

// IsPending 检查申请是否待审批
func (cr *CertificateRequest) IsPending() bool {
return cr.Status == CertificateStatusPending
}

// IsApproved 检查申请是否已批准
func (cr *CertificateRequest) IsApproved() bool {
return cr.Status == CertificateStatusValid
}

// IsRejected 检查申请是否已拒绝
func (cr *CertificateRequest) IsRejected() bool {
return cr.Status == CertificateStatusRejected
}
